src/coreclr/vm/jitinterface.cpp

```diff
@@ -649,6 +649,58 @@ int CEEInfo::getStringLiteral (
     return result;
 }
 
+bool CEEInfo::tryGetNonRandomizedHashCode (
+        CORINFO_MODULE_HANDLE       moduleHnd,
+        mdToken                     metaTOK,
+        bool                        ignoreCase,
+        int*                        pHashCode)
+{
+    CONTRACTL{
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    bool result = false;
+
+    JIT_TO_EE_TRANSITION();
+
+    if (!IsDynamicScope(moduleHnd))
+    {
+        // If ResolveStringRef returns a pinned reference we can return it by value (IAT_VALUE)
+        CORINFO_OBJECT_HANDLE pPinnedString = nullptr;
+        reinterpret_cast<Module*>(moduleHnd)->ResolveStringRef(metaTOK, reinterpret_cast<void**>(&pPinnedString));
+        if (pPinnedString != nullptr)
+        {
+            GCX_COOP();
+            int hashCode;
+
+            DECLARE_ARGHOLDER_ARRAY(args, 1);
+            args[ARGNUM_0] = OBJECTREF_TO_ARGHOLDER(getObjectFromJitHandle(pPinnedString));
+            if (ignoreCase)
+            {
+                // This may trigger ICU loading for non-ASCII output, but it should be fine.
+                // In most cases this is called for an optimized code for a hot block, so, presumably
+                // ICU has already been loaded (e.g by Tier0 code).
```

Comment:
@jkotas it seems that it's beneficial to do for the IgnoreCase, but I'm not sure it's a good idea to trigger ICU here. Perhaps, I should just walk the string and give up if it contains non-ASCII?

Reply:
> give up if it contains non-ASCII?

To extend it to 0xFFFF range without ICU, there is `minipal_toupper_invariant(CHAR16_T)`.

Reply:
I do not think `minipal_toupper_invariant` is guaranteed to produce the exact same result as ICU in all situations (across all ICU versions).

Reply:
For invariant culture, they are updated simultaneously with managed implementation. For invariant case change, managed implementation also has frozen data, which I think is applicable to all platforms alike. (e.g. last update https://github.com/dotnet/runtime/commit/112fa519f2bfdb0f7a2274970dffea2bca4b9d36 modified both `ReadOnlySpan<byte> CategoryCasingLevel1Index` and minipal in one go per the instructions in readme).

Reply:
If invariant culture uses local copy of casing data even when ICU is available (which I think is the case for consistent outputs across platforms, perhaps @tarekgh could confirm this), can we restrict the `0x7F .. 0xFFFF` range optimization to `InvariantCulture{Ignore,}Case` when it is known? Minipal implementation is shared with all runtimes (coreclr, nativeaot and mono); just `#include <minipal/strings.h>`.